tests/rptest/util.py

```diff
@@ -141,6 +143,49 @@ def done():
                err_msg="Segments were not created")
 
 
+def wait_for_removal_of_n_segments(redpanda, topic: str, partition_idx: int,
+                                   n: int,
+                                   original_snapshot: dict[str,
+                                                           list[Segment]]):
+    """
+    Wait until 'n' segments of a partition that are present in the
+    provided snapshot are removed by all brokers.
+
+    :param redpanda: redpanda service used by the test
+    :param topic: topic to wait on
+    :param partition_idx: index of partition to wait on
+    :param n: number of removed segments to wait for
+    :param original_snapshot: snapshot of segments to compare against
+    """
+    def segments_removed():
+        current_snapshot = redpanda.storage(all_nodes=True).segments_by_node(
+            "kafka", topic, 0)
+
+        redpanda.logger.debug(
+            f"Current segment snapshot for topic {topic}: {current_snapshot}")
+
+        # Check how many of the original segments were removed
+        # for each of the nodes in the provided snapshot.
+        for node, original_segs in original_snapshot.items():
+            assert node in current_snapshot
+            current_segs_names = [s.name for s in current_snapshot[node]]
+
+            removed_segments = 0
+            for s in original_segs:
+                if s.name not in current_segs_names:
+                    removed_segments += 1
+
+            if removed_segments < n:
+                return False
+
+        return True
```

Comment:
nit: it sort of looks like this is equivalent to something like `return len(set(orig) - set(removed)) < n`

Reply:
Yeah. I spent more time than I care to admit trying to make this code neat.

To construct a `set`, a [hashable](https://docs.python.org/3/glossary.html#term-hashable) object is required (must implement hash and equality comparison). I thought about it, but couldn't come up with a comparison operator for [Segment](https://github.com/redpanda-data/redpanda/blob/a616818396ca7e0c8d961e8afc1db7774cface61/tests/rptest/services/storage.py#L15). Should the comparison be local, or should it also compare the partition, topic, etc. ?

Maybe the types in storage should all be data-classes and then we get this recursive comparison for free.

Reply:
ahh got it. thanks for the explanation